lib/DxilValidation/DxilContainerValidation.cpp

```diff
@@ -73,6 +78,522 @@ struct DiagRestore {
+
+  ModulePSVSE0.SemanticName = PSVSE0->SemanticName;
+  // Compare all fields.
+  Mismatch |= memcmp(&ModulePSVSE0, PSVSE0, sizeof(PSVSignatureElement0)) != 0;
+  if (Mismatch) {
+    PSVSignatureElement ModulePSVSE(StrTab, IndexTab, &ModulePSVSE0);
+    std::string ModuleStr;
+    raw_string_ostream OS(ModuleStr);
+    ModulePSVSE.Print(OS, SE.GetName(), SemanticIndexVec.data());
+    OS.flush();
+    std::string PartStr;
+    raw_string_ostream OS1(PartStr);
+    PSVSE.Print(OS1, PSVSE.GetSemanticName(), PSVSemanticIndexVec.data());
+    OS1.flush();
+    EmitMismatchError(Name + "Element", PartStr, ModuleStr);
+  }
+}
+
+template <typename T>
+void PSVContentVerifier::VerifyResourceTable(T &ResTab, unsigned &ResourceIndex,
+                                             unsigned PSVVersion) {
+  for (auto &&R : ResTab) {
+    PSVResourceBindInfo1 BI;
+    InitPSVResourceBinding(&BI, &BI, R.get());
+
+    if (PSVVersion > 1) {
+      PSVResourceBindInfo1 *BindInfo =
+          PSV.GetPSVResourceBindInfo1(ResourceIndex);
+      if (memcmp(&BI, BindInfo, sizeof(PSVResourceBindInfo1)) != 0) {
+        std::string ModuleStr = GetDump(BI);
+        std::string PartStr = GetDump(*BindInfo);
+        EmitMismatchError("ResourceBindInfo", PartStr, ModuleStr);
+      }
+    } else {
+      PSVResourceBindInfo0 *BindInfo =
+          PSV.GetPSVResourceBindInfo0(ResourceIndex);
+      if (memcmp(&BI, BindInfo, sizeof(PSVResourceBindInfo0)) != 0) {
+        std::string ModuleStr = GetDump(BI);
+        std::string PartStr = GetDump(*BindInfo);
+        EmitMismatchError("ResourceBindInfo", PartStr, ModuleStr);
+      }
+    }
+    ResourceIndex++;
+  }
+}
+
+void PSVContentVerifier::VerifyResources(unsigned PSVVersion) {
+  UINT uCBuffers = DM.GetCBuffers().size();
+  UINT uSamplers = DM.GetSamplers().size();
+  UINT uSRVs = DM.GetSRVs().size();
+  UINT uUAVs = DM.GetUAVs().size();
+  unsigned ResourceCount = uCBuffers + uSamplers + uSRVs + uUAVs;
+  if (PSV.GetBindCount() != ResourceCount) {
+    EmitMismatchError("ResourceCount", std::to_string(PSV.GetBindCount()),
+                      std::to_string(ResourceCount));
+    return;
+  }
+  // Verify each resource table.
+  unsigned ResIndex = 0;
+  // CBV
+  VerifyResourceTable(DM.GetCBuffers(), ResIndex, PSVVersion);
+  // Sampler
+  VerifyResourceTable(DM.GetSamplers(), ResIndex, PSVVersion);
+  // SRV
+  VerifyResourceTable(DM.GetSRVs(), ResIndex, PSVVersion);
+  // UAV
+  VerifyResourceTable(DM.GetUAVs(), ResIndex, PSVVersion);
+}
+
+void PSVContentVerifier::VerifyEntryProperties(const ShaderModel *SM,
+                                               PSVRuntimeInfo0 *PSV0,
+                                               PSVRuntimeInfo1 *PSV1,
+                                               PSVRuntimeInfo2 *PSV2) {
+  PSVRuntimeInfo3 DMPSV;
+  memset(&DMPSV, 0, sizeof(PSVRuntimeInfo3));
+  hlsl::InitPSVRuntimeInfo(&DMPSV, &DMPSV, &DMPSV, &DMPSV, DM);
```

Comment:
This is a weird usage of a weird function whose previous use has become so mismatched as to be useless. 

First, this function does nothing with the pInfo3 parameter at all. Second, its existing usage in `DxilPSVWriter::Write` makes only one usage of any of the out params and that is to assert that `pInfo1` is not null, which this function doesn't affect. 

I suggest pulling out the old usage entirely and changing this function to your needs where it takes a single pInfo parameter and initializes all the things that are added after below

Reply:
Split into 3 functions.